### ISSUE TITLE
Making horizontal color fade

### ISSUE THREAD

**Issue by diego-rapoport** [qtile/qtile#1856]
Wouldn't it be nice looking to have the possibility of making a color fade horizontally? For now I just can make it from top to bottom and it's limiting. Could be something like an extra argument `background=([color1,color2],'x')`. Default would be y so people don't need to write it all the time.

**Comment by ramnes** [comment]
I'd rather make a `Gradient` object at this point, as we can imagine any direction, any number of colors, shape, etc. PR welcome!

**Comment by diego-rapoport** [comment]
Thank you very much! I'm new to qtile and already loving it. Thanks for all the improvements and your work. I'm sort of new to python, but if there's any help I could give to the project it would be my pleasure.

**Comment by elParaguayo** [comment]
I'm happy to have a go at this as I'm keen to learn how qtile works. I've got a pretty good idea as to how we could implement a Gradient object to handle any direction. 

@ramnes I think Cairo just has linear and radial gradients so I'm not sure about different shapes but I may just start with linear for now!

**Comment by elParaguayo** [comment]
So... I can make a gradient object which can handle any angle. So far, so good...
![gradient](https://user-images.githubusercontent.com/946265/99725934-bd62a700-2aad-11eb-8c9b-08796dee7d1f.png)

The issue I've got is the other widgets draw over it and I can't make them transparent to show the bar underneath.

You can apply custom gradients to individual widgets too.

**Comment by m-col** [comment]
> The issue I've got is the other widgets draw over it and I can't make them transparent to show the bar underneath

Just thinking about ways that this could be resolved, and 2 ideas come to mind:

- give every widget its own window that sits above the bar's window (super hacky... will definitely have Z stacking issues)
- rather than have widgets draw to the window, they could add mask their area over bar's `drawer`, so if the widget has #ffffffff background then it will show the bar's background, but anything else will still be visible. (bit more complicated - might require redrawing the entire bar every time a single widget changes?)

**Comment by elParaguayo** [comment]
Yes - think it's that last point that's the right way to do it. Need to do some research on this.

**Comment by elParaguayo** [comment]
Funnily enough, thinking about this some more, while I do think the last approach is the right one, I suspect that's _way_ beyond my coding skills/understanding. 

I've been thinking about how else to do it, thought I had a good idea, and then realised it was just what @m-col had said in his first bullet!

**Comment by elParaguayo** [comment]
I think the correct answer to this is for qtile to generate 32 bit depth windows which allow for the background to include an alpha channel, rather than using the opacity value.

I think [this post](https://github.com/qtile/qtile/issues/1059#issuecomment-298073985) in an earlier issue was on the right track.

**Comment by elParaguayo** [comment]
This is solved via `GradientDecoration`s in qtile-extras: https://qtile-extras.readthedocs.io/en/stable/manual/ref/decorations.html#gradientdecoration
